R/ensure-source.R

```diff
@@ -1,62 +1,94 @@
-  if (is_url(file)) {
+ensure_source <- function(name) {
+  # if this is a direct file reference, return quickly.
+  if (is_url(name)) {
     # TODO: validate that it exists?
-    return(file)
-  } else if (file.exists(file)) {
+    return(name)
+  } else if (file.exists(name)) {
     # TODO: wrap in some object?
-    return(file)
-  } else if (file.exists(data_file(file))) {
-    return(data_file(file))
+    return(name)
+  }
+
+  # If the source doesn't exist we need to create it
+  # Make sure data dirs exist
+  if (!dir.exists(source_data_file(""))) {
+    dir.create(source_data_file(""))
+  }
+  if (!dir.exists(temp_data_file(""))) {
+    dir.create(temp_data_file(""))
   }
 
-  # Look up, download it
-  known <- known_sources[[file]]
+  known <- known_sources[[name]]
   if (!is.null(known)) {
-    if (!dir.exists(data_file(""))) {
-      # Make sure data dir exists
-      dir.create(data_file(""))
+    filename <- source_filename(name)
+
+    # Check for places this file might already be and return those.
+    if (file.exists(data_file(filename))) {
+      # if the file is in our temp storage or source storage, go for it there.
+      return(data_file(filename))
     }
-    ext <- file_ext(known$url)
-    file <- paste(data_file(file), ext, sep = ".")
+
+    # Look up, download it
+    ext <- file_ext(filename)
+    file <- source_data_file(filename)
     if (!file.exists(file)) {
       # override the timeout
       # TODO: retry with backoff instead of just overriding? or use `curl`?
       with_options(
         new = list(timeout = 600),
         utils::download.file(known$url, file, mode = "wb")
       )
-      # run the post processing only once.
-      on.exit({
-        if (!is.null(known$post_process)) {
-          known$post_process(file)
-        }
-      })
-    }
-    if (ext == "csv.gz") {
-      if (!file.exists(file_with_ext(file, "csv"))) {
-        # This could be done more efficiently
-        # Could shell out to `gunzip` but that assumes the command exists
-        gunzip(file, file_with_ext(file, "csv"), remove = FALSE)
-      }
-      file <- file_with_ext(file, "csv")
     }
+  } else if (!is.null(test_sources[[name]])) {
+    test <- test_sources[[name]]
+    file <- system.file("test_data", test$filename, package = "arrowbench")
   } else {
-    stop(file, " does not exist", call. = FALSE)
+    stop(name, " is not a known source", call. = FALSE)
   }
   file
 }
 
-data_file <- function(..., local_dir = getOption("arrowbench.local_dir", getwd())) {
-  file.path(local_dir, "source_data", ...)
+source_data_file <- function(...) {
+  file.path(local_data_dir(), ...)
+}
+
+temp_data_file <- function(...) {
+  source_data_file("temp", ...)
+}
+
+#' Find a data file
+#'
+#' This looks in both the source dir ("data") as well as the temp directory
+#' ("data/temp") and returns the temp directory file if it exists, if it doesn't
+#' but the source dir exists it will return that, if neither exist it returns
```

Comment:
This behavior is confusing, can you explain why this is desirable?

Reply:
It's more or less to match how the python benchmarks package works. Which is the flow:

* if a data source exists in `data/` use that
* if a data source exists in `data/temp` use that
* if there is no data source in either, you will need to end up needing to create it and put it in `data/temp`

I absolutely agree this function as it is is very odd — I was trying (probably too hard) to be semi-tactical with these changes and not change this internal API too much. But the more that I used it the more that I think there three functions aren't great.

Instead, what I think we should do (and I will do unless you have a better idea) is to have a function that returns the path in `data` or `data/temp` if it exists, and null if it doesn't. And then put the logic for "if this is `NULL` do the things that are needed to make the file exist" in `ensure_format()` (and a bit in `ensure_source()` too probably, though that will be a bit more limited). That way the fallback / search space logic is separated from the where-to-put-it logic and aren't both relying on code that returns paths that sometimes exist and sometimes do not.

Reply:
Yeah I agree that sounds better